src/computedevs.jl

```diff
@@ -0,0 +1,80 @@
+# adaptors for converting abstract arrays to Base.Array
+
+
+"""
+    abstract type AbstractComputingDevice
+
+Supertype for CPU and GPU computing devices.
+
+`adapt(dev::AbstractComputingDevice, x)` adapts `x` for `dev`.
+"""
+abstract type AbstractComputingDevice end
+export AbstractComputingDevice
+
+
+"""
+    struct ComputingDeviceIndependent <: AbstractComputingDevice
+
+`get_computing_device(x) === ComputingDeviceIndependent()` indicates
+that `x` is not tied to a specific computing device. This typically
+means that x is a statically allocated object.
+"""
+struct ComputingDeviceIndependent <: AbstractComputingDevice end
+export ComputingDeviceIndependent
+
+
+"""
+    struct CPUDevice <: AbstractComputingDevice
+
+`CPUDevice()` is the default CPU device.
+"""
+struct CPUDevice <: AbstractComputingDevice end
+export CPUDevice
+
+
+
+"""
+    abstract type AbstractGPUDevice <: AbstractComputingDevice
+
+Supertype for GPU computing devices.
+"""
+abstract type AbstractGPUDevice <: AbstractComputingDevice end
+export AbstractGPUDevice
+
+
+
+const _incompatible_devs = ArgumentError("Incompatible computing devices")
+
+select_computing_device(a::ComputingDeviceIndependent, ::ComputingDeviceIndependent) = a
```

Comment:
I vote for a name like `promote_computing_device`. "Select" makes it sound like you're doing some imperative action to enable some device, whereas it's really just determining promotion rules.

Reply:
> I vote for a name like `promote_computing_device`

I actually thought about it in that term as well.

Reply:
I was tempted to overload `Base.promote` initially, but then thought that would be stretching it too far. :-) I like `promote_computing_device`.

Reply:
Renaming to promote_computing_device